### PR TITLE
Preallocate buffers for FixedSizeBinary array creation

### PR DESCRIPTION
**Pull request by maxburke**
Closes issue #3792

### CODE REVIEW

File: arrow-array/src/array/fixed_size_binary_array.rs [hunk 1]
```diff
@@ -234,8 +242,10 @@ impl FixedSizeBinaryArray {
     {
         let mut len = 0;
         let mut byte = 0;
-        let mut null_buf = MutableBuffer::from_len_zeroed(0);
-        let mut buffer = MutableBuffer::from_len_zeroed(0);
+
+        let iter_size_hint = iter.size_hint().0;
+        let mut null_buf = MutableBuffer::from_len_zeroed((iter_size_hint + 7) / 8);
```
Comment (tustvold) [inline arrow-array/src/array/fixed_size_binary_array.rs:247]: I would perhaps expect to see `MutableBuffer::new` here - https://docs.rs/arrow-buffer/latest/arrow_buffer/struct.MutableBuffer.html#method.new?

File: arrow-array/src/array/fixed_size_binary_array.rs [hunk 2]
```diff
@@ -141,8 +141,11 @@ impl FixedSizeBinaryArray {
         let mut len = 0;
         let mut size = None;
         let mut byte = 0;
-        let mut null_buf = MutableBuffer::from_len_zeroed(0);
+
+        let iter_size_hint = iter.size_hint().0;
+        let mut null_buf = MutableBuffer::from_len_zeroed((iter_size_hint + 7) / 8);
```
Comment (tustvold) [inline arrow-array/src/array/fixed_size_binary_array.rs:146]: Can we use bit_util::ceil here instead

File: arrow-array/src/array/fixed_size_binary_array.rs (hunk 1 above)
Comment (maxburke) [reply arrow-array/src/array/fixed_size_binary_array.rs]: Done

File: arrow-array/src/array/fixed_size_binary_array.rs (hunk 2 above)
Comment (maxburke) [reply arrow-array/src/array/fixed_size_binary_array.rs]: Done

**Review by tustvold — APPROVED** [review]

**Comment by tustvold** [comment]
I wonder if we also need to update try_from_sparse_iter_with_size?

**Pull request merged**